File: ios/runanywhere-sdks-latest/examples/android/RunAnywhereAI/app/src/main/java/com/runanywhere/runanywhereai/presentation/settings/SettingsViewModel.kt
package com.runanywhere.runanywhereai.presentation.settings

import android.app.Application
import android.content.Context
import timber.log.Timber
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.runanywhere.sdk.public.RunAnywhere
import com.runanywhere.sdk.public.events.EventBus
import com.runanywhere.sdk.public.events.ModelEvent
import com.runanywhere.sdk.public.extensions.clearCache
import com.runanywhere.sdk.public.extensions.deleteModel
import com.runanywhere.sdk.public.extensions.storageInfo
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.filterIsInstance
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Simple stored model info for settings display
 */
data class StoredModelInfo(
    val id: String,
    val name: String,
    val size: Long,
)

/**
 * Settings UI State
 */
@OptIn(kotlin.time.ExperimentalTime::class)
data class SettingsUiState(
    // Logging Configuration
    val analyticsLogToLocal: Boolean = false,
    // Storage Overview
    val totalStorageSize: Long = 0L,
    val availableSpace: Long = 0L,
    val modelStorageSize: Long = 0L,
    // Downloaded Models
    val downloadedModels: List<StoredModelInfo> = emptyList(),
    // API Configuration
    val apiKey: String = "",
    val baseURL: String = "",
    val isApiKeyConfigured: Boolean = false,
    val isBaseURLConfigured: Boolean = false,
    val showApiConfigSheet: Boolean = false,
    val showRestartDialog: Boolean = false,
    // Generation Settings
    val temperature: Float = 0.7f,
    val maxTokens: Int = 1000,
    val systemPrompt: String = "",
    // Loading states
    val isLoading: Boolean = false,
    val errorMessage: String? = null,
)

/**
 * Settings ViewModel
 *
 * This ViewModel manages:
 * - Storage overview via RunAnywhere.getStorageInfo()
 * - Model management via RunAnywhere storage APIs
 * - API configuration (API key and base URL)
 */
class SettingsViewModel(application: Application) : AndroidViewModel(application) {
    private val _uiState = MutableStateFlow(SettingsUiState())
    val uiState: StateFlow<SettingsUiState> = _uiState.asStateFlow()

    private val encryptedPrefs by lazy {
        val masterKey = MasterKey.Builder(application)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        EncryptedSharedPreferences.create(
            application,
            ENCRYPTED_PREFS_FILE,
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }

    // Preference file for general app settings (Analytics, etc)
    private val settingsPrefs by lazy {
        application.getSharedPreferences(SETTINGS_PREFS, Context.MODE_PRIVATE)
    }

    // Preference file specifically for LLM generation parameters
    private val generationPrefs by lazy {
        application.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    companion object {
        private const val ENCRYPTED_PREFS_FILE = "runanywhere_secure_prefs"
        private const val SETTINGS_PREFS = "runanywhere_settings"
        private const val KEY_API_KEY = "runanywhere_api_key"
        private const val KEY_BASE_URL = "runanywhere_base_url"
        private const val KEY_DEVICE_REGISTERED = "com.runanywhere.sdk.deviceRegistered"
        private const val KEY_ANALYTICS_LOG_LOCAL = "analyticsLogToLocal"

        // Generation settings constants (match iOS key names)
        private const val PREFS_NAME = "generation_settings"
        private const val KEY_TEMPERATURE = "defaultTemperature"
        private const val KEY_MAX_TOKENS = "defaultMaxTokens"
        private const val KEY_SYSTEM_PROMPT = "defaultSystemPrompt"

        /**
         * Get stored API key (for use at app launch)
         */
        fun getStoredApiKey(context: Context): String? {
            return try {
                val masterKey = MasterKey.Builder(context)
                    .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                    .build()
                val prefs = EncryptedSharedPreferences.create(
                    context,
                    ENCRYPTED_PREFS_FILE,
                    masterKey,
                    EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                    EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
                )
                val value = prefs.getString(KEY_API_KEY, null)
                if (value.isNullOrEmpty()) null else value
            } catch (e: Exception) {
                Timber.e(e, "Failed to get stored API key")
                null
            }
        }

        /**
         * Get stored base URL (for use at app launch)
         * Automatically adds https:// if no scheme is present
         */
        fun getStoredBaseURL(context: Context): String? {
            return try {
                val masterKey = MasterKey.Builder(context)
                    .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                    .build()
                val prefs = EncryptedSharedPreferences.create(
                    context,
                    ENCRYPTED_PREFS_FILE,
                    masterKey,
                    EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                    EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
                )
                val value = prefs.getString(KEY_BASE_URL, null)
                if (value.isNullOrEmpty()) return null

                // Normalize URL by adding https:// if no scheme present
                val trimmed = value.trim()
                if (trimmed.startsWith("http://") || trimmed.startsWith("https://")) {
                    trimmed
                } else {
                    "https://$trimmed"
                }
            } catch (e: Exception) {
                Timber.e(e, "Failed to get stored base URL")
                null
            }
        }

        /**
         * Check if custom configuration is set
         */
        fun hasCustomConfiguration(context: Context): Boolean {
            return getStoredApiKey(context) != null && getStoredBaseURL(context) != null
        }

        /**
         * Data class for generation settings
         */
        data class GenerationSettings(
            val temperature: Float,
            val maxTokens: Int,
            val systemPrompt: String?
        )

        /**
         * Get generation settings (for use by ChatViewModel)
         */
        fun getGenerationSettings(context: Context): GenerationSettings {
            val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            val temperature = prefs.getFloat(KEY_TEMPERATURE, 0.7f)
            val maxTokens = prefs.getInt(KEY_MAX_TOKENS, 1000)
            val systemPrompt = prefs.getString(KEY_SYSTEM_PROMPT, "")

            return GenerationSettings(
                temperature = temperature,
                maxTokens = maxTokens,
                systemPrompt = if (systemPrompt.isNullOrEmpty()) null else systemPrompt
            )
        }
    }

    init {
        loadAnalyticsPreference()
        loadApiConfiguration()
        loadGenerationSettings()
        loadStorageData()
        subscribeToModelEvents()
    }

    private fun loadAnalyticsPreference() {
        val value = settingsPrefs.getBoolean(KEY_ANALYTICS_LOG_LOCAL, false)
        _uiState.update { it.copy(analyticsLogToLocal = value) }
    }

    fun updateAnalyticsLogToLocal(value: Boolean) {
        _uiState.update { it.copy(analyticsLogToLocal = value) }
        settingsPrefs.edit().putBoolean(KEY_ANALYTICS_LOG_LOCAL, value).apply()
    }

    /**
     * Subscribe to SDK model events to automatically refresh storage when models are downloaded/deleted
     */
    private fun subscribeToModelEvents() {
        viewModelScope.launch {
            EventBus.events
                .filterIsInstance<ModelEvent>()
                .collect { event ->
                    when (event.eventType) {
                        ModelEvent.ModelEventType.DOWNLOAD_COMPLETED -> {
                            Timber.d("📥 Model download completed: ${event.modelId}, refreshing storage...")
                            loadStorageData()
                        }
                        ModelEvent.ModelEventType.DELETED -> {
                            Timber.d("🗑️ Model deleted: ${event.modelId}, refreshing storage...")
                            loadStorageData()
                        }
                        else -> {
                            // Other events don't require storage refresh
                        }
                    }
                }
        }
    }

    /**
     * Load storage data using SDK's storageInfo() API
     */
    private fun loadStorageData() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }

            try {
                Timber.d("Loading storage info via storageInfo()...")

                // Use SDK's storageInfo()
                val storageInfo = RunAnywhere.storageInfo()

                // Map stored models to UI model
                val storedModels =
                    storageInfo.storedModels.map { model ->
                        StoredModelInfo(
                            id = model.id,
                            name = model.name,
                            size = model.size,
                        )
                    }

                Timber.d("Storage info received:")
                Timber.d("  - Total space: ${storageInfo.deviceStorage.totalSpace}")
                Timber.d("  - Free space: ${storageInfo.deviceStorage.freeSpace}")
                Timber.d("  - Model storage size: ${storageInfo.totalModelsSize}")
                Timber.d("  - Stored models count: ${storedModels.size}")

                _uiState.update {
                    it.copy(
                        totalStorageSize = storageInfo.deviceStorage.totalSpace,
                        availableSpace = storageInfo.deviceStorage.freeSpace,
                        modelStorageSize = storageInfo.totalModelsSize,
                        downloadedModels = storedModels,
                        isLoading = false,
                    )
                }

                Timber.d("Storage data loaded successfully")
            } catch (e: Exception) {
                Timber.e(e, "Failed to load storage data")
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        errorMessage = "Failed to load storage data: ${e.message}",
                    )
                }
            }
        }
    }

    /**
     * Refresh storage data
     */
    fun refreshStorage() {
        loadStorageData()
    }

    /**
     * Delete a downloaded model
     */
    fun deleteModelById(modelId: String) {
        viewModelScope.launch {
            try {
                Timber.d("Deleting model: $modelId")
                // Use SDK's deleteModel extension function
                RunAnywhere.deleteModel(modelId)
                Timber.d("Model deleted successfully: $modelId")

                // Refresh storage data after deletion
                loadStorageData()
            } catch (e: Exception) {
                Timber.e(e, "Failed to delete model: $modelId")
                _uiState.update {
                    it.copy(errorMessage = "Failed to delete model: ${e.message}")
                }
            }
        }
    }

    /**
     * Clear cache using SDK's clearCache() API
     */
    fun clearCache() {
        viewModelScope.launch {
            try {
                Timber.d("Clearing cache via clearCache()...")
                RunAnywhere.clearCache()
                Timber.d("Cache cleared successfully")

                // Refresh storage data after clearing cache
                loadStorageData()
            } catch (e: Exception) {
                Timber.e(e, "Failed to clear cache")
                _uiState.update {
                    it.copy(errorMessage = "Failed to clear cache: ${e.message}")
                }
            }
        }
    }

    /**
     * Clean temporary files
     */
    fun cleanTempFiles() {
        viewModelScope.launch {
            try {
                Timber.d("Cleaning temp files (via clearing cache)...")
                // Clean temp files by clearing cache
                RunAnywhere.clearCache()
                Timber.d("Temp files cleaned successfully")

                // Refresh storage data after cleaning
                loadStorageData()
            } catch (e: Exception) {
                Timber.e(e, "Failed to clean temp files")
                _uiState.update {
                    it.copy(errorMessage = "Failed to clean temporary files: ${e.message}")
                }
            }
        }
    }

    // ========== Generation Settings Management ==========

    /**
     * Load generation settings from SharedPreferences
     */
    private fun loadGenerationSettings() {
        try {
            val temperature = generationPrefs.getFloat(KEY_TEMPERATURE, 0.7f)
            val maxTokens = generationPrefs.getInt(KEY_MAX_TOKENS, 1000)
            val systemPrompt = generationPrefs.getString(KEY_SYSTEM_PROMPT, "") ?: ""

            _uiState.update {
                it.copy(
                    temperature = temperature,
                    maxTokens = maxTokens,
                    systemPrompt = systemPrompt
                )
            }
            Timber.d("Generation settings loaded - temperature: $temperature, maxTokens: $maxTokens, systemPrompt length: ${systemPrompt.length}")
        } catch (e: Exception) {
            Timber.e(e, "Failed to load generation settings")
        }
    }

    /**
     * Update temperature in UI state
     */
    fun updateTemperature(value: Float) {
        _uiState.update { it.copy(temperature = value) }
    }

    /**
     * Update max tokens in UI state
     */
    fun updateMaxTokens(value: Int) {
        _uiState.update { it.copy(maxTokens = value) }
    }

    /**
     * Update system prompt in UI state
     */
    fun updateSystemPrompt(value: String) {
        _uiState.update { it.copy(systemPrompt = value) }
    }

    /**
     * Save generation settings to SharedPreferences
     */
    fun saveGenerationSettings() {
        viewModelScope.launch {
            try {
                val currentState = _uiState.value
                generationPrefs.edit()
                    .putFloat(KEY_TEMPERATURE, currentState.temperature)
                    .putInt(KEY_MAX_TOKENS, currentState.maxTokens)
                    .putString(KEY_SYSTEM_PROMPT, currentState.systemPrompt)
                    .apply()

                Timber.d("Generation settings saved successfully - temperature: ${currentState.temperature}, maxTokens: ${currentState.maxTokens}")
            } catch (e: Exception) {
                Timber.e(e, "Failed to save generation settings")
                _uiState.update {
                    it.copy(errorMessage = "Failed to save generation settings: ${e.message}")
                }
            }
        }
    }

    // ========== API Configuration Management ==========

    /**
     * Load API configuration from secure storage
     */
    private fun loadApiConfiguration() {
        try {
            val storedApiKey = encryptedPrefs.getString(KEY_API_KEY, "") ?: ""
            val storedBaseURL = encryptedPrefs.getString(KEY_BASE_URL, "") ?: ""

            _uiState.update {
                it.copy(
                    apiKey = storedApiKey,
                    baseURL = storedBaseURL,
                    isApiKeyConfigured = storedApiKey.isNotEmpty(),
                    isBaseURLConfigured = storedBaseURL.isNotEmpty()
                )
            }
            Timber.d("API configuration loaded - apiKey configured: ${storedApiKey.isNotEmpty()}, baseURL configured: ${storedBaseURL.isNotEmpty()}")
        } catch (e: Exception) {
            Timber.e(e, "Failed to load API configuration")
        }
    }

    /**
     * Update API key in UI state
     */
    fun updateApiKey(value: String) {
        _uiState.update { it.copy(apiKey = value) }
    }

    /**
     * Update base URL in UI state
     */
    fun updateBaseURL(value: String) {
        _uiState.update { it.copy(baseURL = value) }
    }

    /**
     * Normalize base URL by adding https:// if no scheme is present
     */
    private fun normalizeBaseURL(url: String): String {
        val trimmed = url.trim()
        if (trimmed.isEmpty()) return trimmed

        return if (trimmed.startsWith("http://") || trimmed.startsWith("https://")) {
            trimmed
        } else {
            "https://$trimmed"
        }
    }

    /**
     * Save API configuration to secure storage
     */
    fun saveApiConfiguration() {
        viewModelScope.launch {
            try {
                val currentState = _uiState.value
                val apiKey = currentState.apiKey
                val normalizedURL = normalizeBaseURL(currentState.baseURL)

                encryptedPrefs.edit()
                    .putString(KEY_API_KEY, apiKey)
                    .putString(KEY_BASE_URL, normalizedURL)
                    .apply()

                _uiState.update {
                    it.copy(
                        baseURL = normalizedURL,
                        isApiKeyConfigured = apiKey.isNotEmpty(),
                        isBaseURLConfigured = normalizedURL.isNotEmpty(),
                        showApiConfigSheet = false,
                        showRestartDialog = true
                    )
                }

                Timber.d("API configuration saved successfully")
            } catch (e: Exception) {
                Timber.e(e, "Failed to save API configuration")
                _uiState.update {
                    it.copy(errorMessage = "Failed to save API configuration: ${e.message}")
                }
            }
        }
    }

    /**
     * Clear API configuration from secure storage
     */
    fun clearApiConfiguration() {
        viewModelScope.launch {
            try {
                encryptedPrefs.edit()
                    .remove(KEY_API_KEY)
                    .remove(KEY_BASE_URL)
                    .apply()

                // Also clear device registration so it re-registers with new config
                clearDeviceRegistration()

                _uiState.update {
                    it.copy(
                        apiKey = "",
                        baseURL = "",
                        isApiKeyConfigured = false,
                        isBaseURLConfigured = false,
                        showRestartDialog = true
                    )
                }

                Timber.d("API configuration cleared successfully")
            } catch (e: Exception) {
                Timber.e(e, "Failed to clear API configuration")
                _uiState.update {
                    it.copy(errorMessage = "Failed to clear API configuration: ${e.message}")
                }
            }
        }
    }

    /**
     * Clear device registration status (forces re-registration on next launch)
     */
    private fun clearDeviceRegistration() {
        val context = getApplication<Application>()
        context.getSharedPreferences("runanywhere_sdk", Context.MODE_PRIVATE)
            .edit()
            .remove(KEY_DEVICE_REGISTERED)
            .apply()
        Timber.d("Device registration cleared - will re-register on next launch")
    }

    /**
     * Show the API configuration sheet
     */
    fun showApiConfigSheet() {
        _uiState.update { it.copy(showApiConfigSheet = true) }
    }

    /**
     * Hide the API configuration sheet
     */
    fun hideApiConfigSheet() {
        // Reload saved configuration when canceling
        loadApiConfiguration()
        _uiState.update { it.copy(showApiConfigSheet = false) }
    }

    /**
     * Dismiss the restart dialog
     */
    fun dismissRestartDialog() {
        _uiState.update { it.copy(showRestartDialog = false) }
    }

    /**
     * Check if API configuration is complete (both key and URL set)
     */
    fun isApiConfigurationComplete(): Boolean {
        val state = _uiState.value
        return state.isApiKeyConfigured && state.isBaseURLConfigured
    }
}

File: ios/runanywhere-sdks-latest/examples/ios/RunAnywhereAI/RunAnywhereAI/Features/Settings/SettingsViewModel.swift
//
//  SettingsViewModel.swift
//  RunAnywhereAI
//
//  Centralized ViewModel for all Settings functionality
//  Follows MVVM pattern - all business logic is here
//

import Foundation
import SwiftUI
import RunAnywhere
import Combine

@MainActor
class SettingsViewModel: ObservableObject {
    // MARK: - Published Properties

    // Generation Settings
    @Published var temperature: Double = 0.7
    @Published var maxTokens: Int = 10000
    @Published var systemPrompt: String = ""

    // API Configuration
    @Published var apiKey: String = ""
    @Published var baseURL: String = ""
    @Published var isApiKeyConfigured: Bool = false
    @Published var isBaseURLConfigured: Bool = false

    // Logging Configuration
    @Published var analyticsLogToLocal: Bool = false

    // Storage Overview
    @Published var totalStorageSize: Int64 = 0
    @Published var availableSpace: Int64 = 0
    @Published var modelStorageSize: Int64 = 0
    @Published var storedModels: [StoredModel] = []

    // UI State
    @Published var showApiKeyEntry: Bool = false
    @Published var isLoadingStorage: Bool = false
    @Published var errorMessage: String?
    @Published var showRestartAlert: Bool = false

    // MARK: - Private Properties

    private var cancellables = Set<AnyCancellable>()
    private let keychainService = KeychainService.shared
    private let apiKeyStorageKey = "runanywhere_api_key"
    private let baseURLStorageKey = "runanywhere_base_url"
    private let temperatureDefaultsKey = "defaultTemperature"
    private let maxTokensDefaultsKey = "defaultMaxTokens"
    private let systemPromptDefaultsKey = "defaultSystemPrompt"
    private let analyticsLogKey = "analyticsLogToLocal"
    private let deviceRegisteredKey = "com.runanywhere.sdk.deviceRegistered"

    // MARK: - Static helpers for app initialization
    static let shared = SettingsViewModel()

    /// Get stored API key (for use at app launch)
    static func getStoredApiKey() -> String? {
        guard let data = try? KeychainService.shared.retrieve(key: "runanywhere_api_key"),
              let value = String(data: data, encoding: .utf8),
              !value.isEmpty else {
            return nil
        }
        return value
    }

    /// Get stored base URL (for use at app launch)
    /// Automatically adds https:// if no scheme is present
    static func getStoredBaseURL() -> String? {
        guard let data = try? KeychainService.shared.retrieve(key: "runanywhere_base_url"),
              let value = String(data: data, encoding: .utf8),
              !value.isEmpty else {
            return nil
        }
        // Normalize URL by adding https:// if no scheme present
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.hasPrefix("http://") || trimmed.hasPrefix("https://") {
            return trimmed
        }
        return "https://\(trimmed)"
    }

    /// Check if custom configuration is set
    static var hasCustomConfiguration: Bool {
        getStoredApiKey() != nil && getStoredBaseURL() != nil
    }

    // MARK: - Initialization

    init() {
        loadSettings()
        setupObservers()
    }

    // MARK: - Setup

    private func setupObservers() {
        // Auto-save temperature changes
        $temperature
            .debounce(for: 0.5, scheduler: DispatchQueue.main)
            .dropFirst() // Skip initial value to avoid saving on init
            .sink { [weak self] newValue in
                self?.saveTemperature(newValue)
            }
            .store(in: &cancellables)

        // Auto-save max tokens changes
        $maxTokens
            .debounce(for: 0.5, scheduler: DispatchQueue.main)
            .dropFirst() // Skip initial value to avoid saving on init
            .sink { [weak self] newValue in
                self?.saveMaxTokens(newValue)
            }
            .store(in: &cancellables)

        // Auto-save system prompt changes
        $systemPrompt
            .debounce(for: 0.5, scheduler: DispatchQueue.main)
            .dropFirst() // Skip initial value to avoid saving on init
            .sink { [weak self] newValue in
                self?.saveSystemPrompt(newValue)
            }
            .store(in: &cancellables)

        // Auto-save analytics logging preference
        $analyticsLogToLocal
            .dropFirst() // Skip initial value to avoid saving on init
            .sink { [weak self] newValue in
                self?.saveAnalyticsLogPreference(newValue)
            }
            .store(in: &cancellables)
    }

    // MARK: - Settings Management

    /// Load all settings from storage
    func loadSettings() {
        loadGenerationSettings()
        loadApiKeyConfiguration()
        loadLoggingConfiguration()
    }

    private func loadGenerationSettings() {
        // Load temperature
        let savedTemperature = UserDefaults.standard.double(forKey: temperatureDefaultsKey)
        temperature = savedTemperature > 0 ? savedTemperature : 0.7

        // Load max tokens
        let savedMaxTokens = UserDefaults.standard.integer(forKey: maxTokensDefaultsKey)
        maxTokens = savedMaxTokens > 0 ? savedMaxTokens : 10000

        // Load system prompt
        systemPrompt = UserDefaults.standard.string(forKey: systemPromptDefaultsKey) ?? ""
    }

    private func loadApiKeyConfiguration() {
        // Load API key from keychain
        if let apiKeyData = try? keychainService.retrieve(key: apiKeyStorageKey),
           let savedApiKey = String(data: apiKeyData, encoding: .utf8),
           !savedApiKey.isEmpty {
            apiKey = savedApiKey
            isApiKeyConfigured = true
        } else {
            apiKey = ""
            isApiKeyConfigured = false
        }

        // Load Base URL from keychain
        if let baseURLData = try? keychainService.retrieve(key: baseURLStorageKey),
           let savedBaseURL = String(data: baseURLData, encoding: .utf8),
           !savedBaseURL.isEmpty {
            baseURL = savedBaseURL
            isBaseURLConfigured = true
        } else {
            baseURL = ""
            isBaseURLConfigured = false
        }
    }

    private func loadLoggingConfiguration() {
        analyticsLogToLocal = keychainService.loadBool(key: analyticsLogKey, defaultValue: false)
    }

    // MARK: - Generation Settings

    private func saveTemperature(_ value: Double) {
        UserDefaults.standard.set(value, forKey: temperatureDefaultsKey)
        print("Settings: Saved temperature: \(value)")
    }

    private func saveMaxTokens(_ value: Int) {
        UserDefaults.standard.set(value, forKey: maxTokensDefaultsKey)
        print("Settings: Saved max tokens: \(value)")
    }

    private func saveSystemPrompt(_ value: String) {
        UserDefaults.standard.set(value, forKey: systemPromptDefaultsKey)
        print("Settings: Saved system prompt (\(value.count) chars)")
    }

    /// Get current generation configuration for SDK usage
    func getGenerationConfiguration() -> GenerationConfiguration {
        GenerationConfiguration(
            temperature: temperature,
            maxTokens: maxTokens,
            systemPrompt: systemPrompt.isEmpty ? nil : systemPrompt
        )
    }

    // MARK: - API Configuration Management

    /// Normalize base URL by adding https:// if no scheme is present
    private func normalizeBaseURL(_ url: String) -> String {
        let trimmed = url.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty {
            return trimmed
        }

        // Check if URL already has a scheme
        if trimmed.hasPrefix("http://") || trimmed.hasPrefix("https://") {
            return trimmed
        }

        // Add https:// prefix
        return "https://\(trimmed)"
    }

    /// Save API key and Base URL to secure storage
    func saveApiConfiguration() {
        var hasError = false

        // Save API key if provided
        if !apiKey.isEmpty {
            if let apiKeyData = apiKey.data(using: .utf8) {
                do {
                    try keychainService.save(key: apiKeyStorageKey, data: apiKeyData)
                    isApiKeyConfigured = true
                    print("Settings: API key saved successfully")
                } catch {
                    errorMessage = "Failed to save API key: \(error.localizedDescription)"
                    hasError = true
                }
            }
        }

        // Save Base URL if provided (normalize to add https:// if missing)
        if !baseURL.isEmpty {
            let normalizedURL = normalizeBaseURL(baseURL)
            baseURL = normalizedURL  // Update the displayed value too

            if let baseURLData = normalizedURL.data(using: .utf8) {
                do {
                    try keychainService.save(key: baseURLStorageKey, data: baseURLData)
                    isBaseURLConfigured = true
                    print("Settings: Base URL saved successfully: \(normalizedURL)")
                } catch {
                    errorMessage = "Failed to save Base URL: \(error.localizedDescription)"
                    hasError = true
                }
            }
        }

        if !hasError {
            showApiKeyEntry = false
            errorMessage = nil
            // Show restart alert
            showRestartAlert = true
        }
    }

    /// Delete API configuration from secure storage
    func clearApiConfiguration() {
        do {
            try keychainService.delete(key: apiKeyStorageKey)
            try keychainService.delete(key: baseURLStorageKey)
            apiKey = ""
            baseURL = ""
            isApiKeyConfigured = false
            isBaseURLConfigured = false
            errorMessage = nil

            // Also clear device registration so it re-registers with new config
            clearDeviceRegistration()

            print("Settings: API configuration cleared successfully")
            showRestartAlert = true
        } catch {
            errorMessage = "Failed to clear API configuration: \(error.localizedDescription)"
        }
    }

    /// Clear device registration status (forces re-registration on next launch)
    func clearDeviceRegistration() {
        UserDefaults.standard.removeObject(forKey: deviceRegisteredKey)
        print("Settings: Device registration cleared - will re-register on next launch")
    }

    /// Show the API configuration sheet
    func showApiKeySheet() {
        showApiKeyEntry = true
    }

    /// Cancel API key entry
    func cancelApiKeyEntry() {
        // Reload the saved configuration if canceling
        loadApiKeyConfiguration()
        showApiKeyEntry = false
    }

    /// Check if API configuration is complete (both key and URL set)
    var isApiConfigurationComplete: Bool {
        isApiKeyConfigured && isBaseURLConfigured
    }

    // MARK: - Logging Configuration

    private func saveAnalyticsLogPreference(_ value: Bool) {
        try? keychainService.saveBool(key: analyticsLogKey, value: value)
        print("Settings: Analytics logging set to: \(value)")
    }

    // MARK: - Storage Management

    /// Load storage information
    func loadStorageData() async {
        isLoadingStorage = true
        errorMessage = nil

        do {
            let storageInfo = await RunAnywhere.getStorageInfo()

            totalStorageSize = storageInfo.appStorage.totalSize
            availableSpace = storageInfo.deviceStorage.freeSpace
            modelStorageSize = storageInfo.totalModelsSize
            storedModels = storageInfo.storedModels

            print("Settings: Loaded storage data - Total: \(totalStorageSize), Available: \(availableSpace)")
        } catch {
            errorMessage = "Failed to load storage data: \(error.localizedDescription)"
        }

        isLoadingStorage = false
    }

    /// Refresh storage information
    func refreshStorageData() async {
        await loadStorageData()
    }

    /// Clear cache
    func clearCache() async {
        do {
            try await RunAnywhere.clearCache()
            await refreshStorageData()
            print("Settings: Cache cleared successfully")
        } catch {
            errorMessage = "Failed to clear cache: \(error.localizedDescription)"
        }
    }

    /// Clean temporary files
    func cleanTempFiles() async {
        do {
            try await RunAnywhere.cleanTempFiles()
            await refreshStorageData()
            print("Settings: Temporary files cleaned successfully")
        } catch {
            errorMessage = "Failed to clean temporary files: \(error.localizedDescription)"
        }
    }

    /// Delete a stored model
    func deleteModel(_ model: StoredModel) async {
        guard let framework = model.framework else {
            errorMessage = "Cannot delete model: unknown framework"
            return
        }

        do {
            try await RunAnywhere.deleteStoredModel(model.id, framework: framework)
            await refreshStorageData()
            print("Settings: Model \(model.name) deleted successfully")
        } catch {
            errorMessage = "Failed to delete model: \(error.localizedDescription)"
        }
    }

    // MARK: - Helper Methods

    /// Format bytes to human-readable string
    func formatBytes(_ bytes: Int64) -> String {
        ByteCountFormatter.string(fromByteCount: bytes, countStyle: .file)
    }

    /// Format bytes to memory string
    func formatMemory(_ bytes: Int64) -> String {
        ByteCountFormatter.string(fromByteCount: bytes, countStyle: .memory)
    }

    /// Check if storage data is available
    var hasStorageData: Bool {
        totalStorageSize > 0
    }

    /// Get storage usage percentage
    var storageUsagePercentage: Double {
        guard availableSpace > 0 else { return 0 }
        let totalDevice = totalStorageSize + availableSpace
        return Double(totalStorageSize) / Double(totalDevice)
    }
}

// MARK: - Supporting Types

struct GenerationConfiguration {
    let temperature: Double
    let maxTokens: Int
    let systemPrompt: String?
}
